server/src/main/java/io/crate/types/DataTypes.java

```diff
@@ -610,6 +610,19 @@ public static boolean isCompatibleType(List<DataType<?>> left, List<DataType<?>>
         return true;
     }
 
+    public static DataType<?> merge(DataType<?> leftType, DataType<?> rightType) {
+        DataType<?> type;
+        if (leftType.id() == ObjectType.ID) {
+            type = ObjectType.merge((ObjectType) leftType, (ObjectType) rightType);
+        } else if (leftType.id() == ArrayType.ID) {
+            type = new ArrayType<>(
+                merge(((ArrayType<?>) leftType).innerType(), ((ArrayType<?>) rightType).innerType()));
+        } else {
+            type = leftType.isConvertableTo(rightType, false) ? rightType : leftType;
```

Comment:
Shouldn't this also check the precedence?

Reply:
Thanks, this helped me realize a bug that slipped my mind. I am working on a [PR](https://github.com/crate/crate/pull/14723) for it and move `union` related commit to it.